pkg/appliance/functions.go

```diff
@@ -776,3 +788,251 @@ func PrettyBytes(v float64) string {
+	}
+
+	go func(wg *sync.WaitGroup, errChan chan error) {
+		wg.Wait()
+		close(errChan)
+	}(&wg, errChan)
+
+	var errs *multierror.Error
+	for err := range errChan {
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		}
+	}
+	log.Info("docker layers downloaded")
+
+	return archive, fileStat, errs.ErrorOrNil()
+}
+
+func downloadDockerImageBundle(ctx context.Context, out io.Writer, client *http.Client, ciMode bool, zipWriter *zip.Writer, registry, image, tag string, p *tui.Progress, mainErrChan chan<- error, wg *sync.WaitGroup) {
+	defer wg.Done()
+	headers := map[string]string{
+		"Accept":       "application/vnd.docker.distribution.manifest.v2+json",
+		"Content-Type": "application/json",
+	}
+
+	// Download mainfest.json
+	manifestURL := fmt.Sprintf("%s/v2/%s/manifests/%s", registry, image, tag)
+	manifestReq, err := http.NewRequestWithContext(ctx, http.MethodGet, manifestURL, nil)
+	if err != nil {
+		mainErrChan <- err
+		return
+	}
+	for k, v := range headers {
+		manifestReq.Header.Set(k, v)
+	}
+	manifestRes, err := client.Do(manifestReq)
+	if manifestRes.StatusCode != http.StatusOK || err != nil {
+		mainErrChan <- api.HTTPErrorResponse(manifestRes, err)
+		return
+	}
+	defer manifestRes.Body.Close()
+
+	manifestBuf := &bytes.Buffer{}
+	if _, err := io.Copy(manifestBuf, manifestRes.Body); err != nil {
+		mainErrChan <- err
+		return
+	}
+	JSONManifest := DockerManifest{}
+	if err := json.Unmarshal(manifestBuf.Bytes(), &JSONManifest); err != nil {
+		mainErrChan <- err
+		return
+	}
+
+	// Download config.json
+	ImageID := JSONManifest.Config.Digest
+	configURL := fmt.Sprintf("%s/v2/%s/blobs/%s", registry, image, ImageID)
+	configReq, err := http.NewRequestWithContext(ctx, http.MethodGet, configURL, nil)
+	if err != nil {
+		mainErrChan <- err
+		return
+	}
+	for k, v := range headers {
+		configReq.Header.Set(k, v)
+	}
+	configRes, err := client.Do(configReq)
+	if configRes.StatusCode != http.StatusOK || err != nil {
+		mainErrChan <- api.HTTPErrorResponse(configRes, err)
+		return
+	}
+	defer configRes.Body.Close()
+
+	configBuf := &bytes.Buffer{}
+	if _, err := io.Copy(configBuf, configRes.Body); err != nil {
+		mainErrChan <- err
+		return
+	}
+
+	// Download layers
+	type bufName struct {
+		name string
+		buf  *bytes.Buffer
+	}
+	layerCount := len(JSONManifest.Layers)
+	layers := make(map[string]io.Reader)
+	bufChan := make(chan bufName)
+	errChan := make(chan error)
+	for _, l := range JSONManifest.Layers {
+		go func(l ManifestConfig, bufChan chan bufName, errChan chan error) {
+			f := strings.Replace(l.Digest, "sha256:", "", 1) + ".tar.gz"
+			layerURL := fmt.Sprintf("%s/v2/%s/blobs/%s", registry, image, l.Digest)
+			layerReq, err := http.NewRequestWithContext(ctx, http.MethodGet, layerURL, nil)
+			if err != nil {
+				errChan <- err
+				return
+			}
+			layerRes, err := client.Do(layerReq)
+			if layerRes.StatusCode != http.StatusOK || err != nil {
```

Comment:
```suggestion
			if err != nil {
```
layerRes can be null, avoid panic by only checking err